### DISCUSSION TITLE
RFC: envd CLI design

### DISCUSSION THREAD

**Discussion by gaocegege** [tensorchord/envd#1279] [Design]
I propose to change the CLI:

- Add a new command `login`, which is a hidden command now
- Rename `envd run` to `envd exec`, to spawn a command installed into the environment.
- Add a new command `envd run`, which is a hidden command `envd create`
- reorder the command categories

```
 Basic Commands:
  init, i           Initializes the current directory with the build.envd file
  build, b          Build the envd environment
  destroy, down, d  Destroy the envd environment
  up, u             Build and run the envd environment
  run               run the envd environment from an existing image
 Expert Commands:
  pause, p   Pause the envd environment
  resume, r  Resume the envd environment
  exec       Spawns a command installed into the environment.
 Management Commands:
  login 
  context           Manage envd contexts
  images, image     Manage envd images
  envs, env, e      Manage envd environments
  top               Show statistics about the containers managed by the environment.
 Admin Commands:
  bootstrap   Bootstrap the envd installation
  completion  Install shell completion scripts for envd
  prune       Clean up the build cache
 Other Commands:
  debug, b    Debug commands
  version, v  Print envd version information
```

**Comment by kemingy** [comment]
* shall we use `envd auth` or `envd login`
* others LGTM

**Reply by gaocegege** [reply]
cc @VoVAllen 

auth LGTM

**Reply by VoVAllen** [reply]
I think login is better. Also you mentioned `envd server login`, what do you think about it?

**Reply by xieydd** [reply]
`auth` lgtm, `login` intuitively not a management command.

**Reply by gaocegege** [reply]
> I think login is better. Also, you mentioned `envd server login`, what do you think about it?

Yep, I was thinking if we can expose some admin commands for server if the user is admin.

For example, `envd server users` to list all users. `envd server gpus` to list all GPU infos.

**Reply by kemingy** [reply]
I think we don't have to use CLI for everything. Especially when you need to check the audit log. The dashboard should be the suitable way.

**Reply by gaocegege** [reply]
Yep, thus I do not propose `envd server login`. Because I do not see too many commands in the `server` subcommand category.

**Reply by gaocegege** [reply]
> I think login is better. Also you mentioned `envd server login`, what do you think about it?

Could you please explain why login is better from your side?

**Reply by VoVAllen** [reply]
Because terraform and dagger use `login`. I think `login` is more common.

- https://developer.hashicorp.com/terraform/cli/commands/login

**Reply by gaocegege** [reply]
SGTM.

**Comment by gaocegege** [comment]
Besides this, I suggest adding `envd attach`, to attach to the existing env.

**Reply by xieydd** [reply]
candidate: `envd switch`

**Reply by gaocegege** [reply]
> candidate: `envd switch`

I did not see other tools use switch to attach to the env. Could you please give some examples?

**Reply by gaocegege** [reply]
Some candidates:

- envd ssh
- envd shell

**Reply by xieydd** [reply]
I would like to know `attach to the existing env`  means  you need to switch to other enviroment.Maybe i misunderstood.

**Reply by gaocegege** [reply]
This command is to attach from the host OS to the container.

**Reply by gaocegege** [reply]
After thinking, prefer ssh since users may know there is a ssh connection underlying.

**Reply by xieydd** [reply]
> This command is to attach from the host OS to the container.

`ssh` ltgm.
